sdk/examples/xo_python/sawtooth_xo/processor/xo_state.py

```diff
@@ -0,0 +1,175 @@
+# limitations under the License.
+# -----------------------------------------------------------------------------
+
+import hashlib
+
+from sawtooth_sdk.processor.exceptions import InternalError
+
+
+XO_NAMESPACE = hashlib.sha512('xo'.encode("utf-8")).hexdigest()[0:6]
+
+
+def _make_xo_address(name):
+    return XO_NAMESPACE + \
+        hashlib.sha512(name.encode('utf-8')).hexdigest()[:64]
+
+
+class Game(object):
+    def __init__(self, name, board, state, player1, player2):
+        self.name = name
+        self.board = board
+        self.state = state
+        self.player1 = player1
+        self.player2 = player2
+
+
+class XoState(object):
+
+    TIMEOUT = 3
+
+    def __init__(self, context):
+        """Constructor.
+
+        Args:
+            context (sawtooth_sdk.processor.context.Context): Access to
+                validator state from within the transaction processor.
+        """
+
+        self._context = context
+        self._address_cache = {}
+
+    def delete_game(self, game_name):
+        """Delete the Game named game_name from state.
+
+        Args:
+            game_name (str): The name.
+
+        Raises:
+            KeyError: The Game with game_name does not exist.
+        """
+
+        address = _make_xo_address(game_name)
+        self._context.delete_state(
+            [address],
+            timeout=self.TIMEOUT)
+
+        games = self._load_games(game_name=game_name)
+
+        del games[game_name]
+        if games:
+            self._store_game(game_name, games=games)
+        else:
+            self._delete_game(game_name)
+
+    def set_game(self, game_name, game):
+        """Store the game in the validator state.
+
+        Args:
+            game_name (str): The name.
+            game (Game): The information specifying the current game.
+        """
+
+        games = self._load_games(game_name=game_name)
+
+        games[game_name] = game
+
+        self._store_game(game_name, games=games)
+
+    def get_game(self, game_name):
+        """Get the game associated with game_name.
+
+        Args:
+            game_name (str): The name.
+
+        Returns:
+            (Game): All the information specifying a game.
+        """
+
+        return self._load_games(game_name=game_name).get(game_name)
+
+    def _store_game(self, game_name, games):
+        address = _make_xo_address(game_name)
+
+        state_data = self._serialize(games)
+
+        self._context.set_state(
+            {address: state_data},
+            timeout=self.TIMEOUT)
+
+    def _delete_game(self, game_name):
+        address = _make_xo_address(game_name)
+
+        self._context.delete_state(
+            [address],
+            timeout=self.TIMEOUT)
+
+    def _load_games(self, game_name):
+        address = _make_xo_address(game_name)
+
+        if address in self._address_cache:
+            serialized_games = self._address_cache[address]
+            games = self._deserialize(serialized_games)
+        else:
```

Comment:
When you are getting/setting/deleting state the address cache needs to be updated to accurately reflect whatever the validator has, otherwise the address cache will cause problems or not be useful.